File: personalizatio-sdk/src/main/java/com/personalizatio/stories/views/storyItem/TextBlockView.java
package com.personalizatio.stories.views.storyItem;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.View;
import android.widget.FrameLayout;

import androidx.annotation.FontRes;
import androidx.annotation.NonNull;
import androidx.core.content.res.ResourcesCompat;
import com.personalizatio.R;
import com.personalizatio.stories.models.elements.TextBlockElement;
import com.personalizatio.utils.ViewUtils;

final public class TextBlockView extends androidx.appcompat.widget.AppCompatTextView {

	public TextBlockView(@NonNull Context context) {
		super(context);
	}

	@SuppressLint("ResourceAsColor")
	public void updateView(TextBlockElement element, int parentHeight, int parentTopOffset) {
		var layoutParams = new FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.WRAP_CONTENT);
		setLayoutParams(layoutParams);

		setText(element.getTextInput());

		var fontSize = element.getFontSize();

		setPadding(0, fontSize, 0, fontSize);

		setY(parentHeight * element.getYOffset() / 100f + parentTopOffset);

		setTextSize(fontSize);

		var typeface = ResourcesCompat.getFont(getContext(), getFontRes(element.getFontType(), element.isBold(), element.isItalic()));
		setTypeface(typeface, ViewUtils.getTypefaceStyle(element.isBold(), element.isItalic()));

		setTextAlignment(getTextAlignment(element.getTextAlign()));

		setLineSpacing(getLineHeight(), (float)element.getTextLineSpacing());

		setBackgroundColor(element.getTextBackgroundColor(), element.getTextBackgroundColorOpacity());
		ViewUtils.setTextColor(getContext(), this, element.getTextColor(), R.color.white);
	}

	private @FontRes int getFontRes(String fontType, boolean bold, boolean italic) {
		switch (fontType) {
            case "serif": {
				if (bold && italic) return R.font.droid_serif_bold_italic;
				if (bold) return R.font.droid_serif_bold;
				if (italic) return R.font.droid_serif_italic;
				return R.font.droid_serif_regular;
			}
			case "sans-serif": {
				if (bold) return R.font.droid_sans_bold;
				return R.font.droid_sans_regular;
			}
			case "monospaced":
			default: {
				return R.font.droid_sans_mono;
			}
		}
	}

	private static int getTextAlignment(String textAlign) {
		return switch (textAlign) {
			case "center" -> View.TEXT_ALIGNMENT_CENTER;
			case "right" -> View.TEXT_ALIGNMENT_TEXT_END;
			default -> View.TEXT_ALIGNMENT_TEXT_START;
		};
	}

	@SuppressLint("ResourceAsColor")
	private void setBackgroundColor(String colorString, String colorOpacityString) {
		if( !colorString.startsWith("#") ) {
			colorString = "#FFFFFF";
		}

		var colorOpacity = GetColorOpacity(colorOpacityString);
		var colorOpacityValueString = Integer.toString(colorOpacity,16);
		if( colorOpacityValueString.length() == 1) colorOpacityValueString = 0 + colorOpacityValueString;

		var fullColorString = colorString.replace("#", "#" + colorOpacityValueString);

		var color = ViewUtils.getColor(getContext(), fullColorString, R.color.white);
		setBackgroundColor(color);
	}

	private int GetColorOpacity(String percentsString) {
		var percents = 0;

		try {
			percents = Integer.parseInt(percentsString);
		}
		catch( NumberFormatException e ) {
			try {
				if (!percentsString.isEmpty()) {
					percents = Integer.parseInt(percentsString.substring(0, percentsString.length() - 1));
				}
			}
			catch( NumberFormatException ignored) {
			}
		}

		return ViewUtils.MAX_COLOR_CHANNEL_VALUE * percents / 100;
	}
}

File: personalizatio-sdk/src/main/kotlin/com/personalizatio/stories/views/storyItem/TextBlockView.kt
package com.personalizatio.stories.views.storyItem

import android.annotation.SuppressLint
import android.content.Context
import android.widget.FrameLayout
import androidx.appcompat.widget.AppCompatTextView
import com.personalizatio.R
import com.personalizatio.stories.models.elements.TextBlockElement
import com.personalizatio.utils.ViewUtils

class TextBlockView(context: Context) : AppCompatTextView(context) {

    @SuppressLint("ResourceAsColor")
    fun updateView(element: TextBlockElement, parentHeight: Int, parentTopOffset: Int) {
        val layoutParams = FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.WRAP_CONTENT)
        setLayoutParams(layoutParams)

        text = element.textInput
        setPadding(0, element.fontSize, 0, element.fontSize)
        y = parentHeight * element.yOffset / 100f + parentTopOffset
        textSize = element.fontSize.toFloat()

        val typeface = ResourcesCompat.getFont(
            context,
            ViewUtils.getFontRes(element.fontType, element.isBold, element.isItalic)
        )
        setTypeface(typeface, ViewUtils.getTypefaceStyle(element.isBold, element.isItalic))

        textAlignment = getTextAlignment(element.textAlign)
        setLineSpacing(lineSpacingExtra, element.textLineSpacing.toFloat())

        val colorOpacity = ViewUtils.getColorOpacity(element.textBackgroundColorOpacity)
        if (colorOpacity > 0) {
            val paddingPx = ViewUtils.dpToPx(PADDING_DP.toFloat(), context).toInt()
            setPadding(paddingPx, paddingPx, paddingPx, paddingPx)
        }

        ViewUtils.setBackgroundColor(this, element.textBackgroundColor, colorOpacity, context)
        ViewUtils.setTextColor(context, this, element.textColor, R.color.white)
    }

    companion object {
        private const val PADDING_DP = 16

        private fun getTextAlignment(textAlign: String): Int {
            return when (textAlign) {
                "center" -> TEXT_ALIGNMENT_CENTER
                "right" -> TEXT_ALIGNMENT_TEXT_END
                else -> TEXT_ALIGNMENT_TEXT_START
            }
        }
    }
}
